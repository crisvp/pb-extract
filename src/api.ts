import { BaseAuthStore, Client } from '@crisvp/pocketbase-js';

import { CollectionDescription, CollectionDescriptionRaw } from './types';
import { normalizeDescriptions } from './translations.js';

export interface ConnectOptions {
  authStore?: BaseAuthStore;
  lang?: string;
}

export interface ClientOptions {
  adminUser: string;
  adminPassword: string;
}

export async function connectDatabase(
  url: string = 'http://127.0.0.1:8090/',
  adminUser: string,
  adminPassword: string,
  connectOptions: ConnectOptions = {}
): Promise<Client> {
  const opts = {
    ...connectOptions,
  };
  const pb = new Client(url, opts.authStore, opts.lang);

  try {
    await pb.admins.authWithPassword(adminUser, adminPassword);
  } catch (e) {
    if (e.status === 401) throw new Error('Failed to authenticate admin user');
    else throw e;
  }
  return pb;
}

async function collectionsTable(pb: Client): Promise<CollectionDescriptionRaw[]> {
  const allCollections = (await pb.collections.getFullList()) as unknown as CollectionDescriptionRaw[];
  const collections = allCollections.filter(r => !r.name.startsWith('_'));

  return collections;
}

export async function readCollections(
  url: string,
  { adminUser, adminPassword }: ClientOptions
): Promise<CollectionDescription[]> {
  const pb = await connectDatabase(url, adminUser, adminPassword);
  const collections = await collectionsTable(pb);
  const normalizedCollections = normalizeDescriptions(collections);
  return normalizedCollections;
}
