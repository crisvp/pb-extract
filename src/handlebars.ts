import handlebars from 'handlebars';

/**
 * Handlebars helper to convert a string to PascalCase. Pass any number of arguments.
 *
 * @param {...any[]} args E.g., use: {{pascalCase "hello" "world"}}
 * @returns {string}
 */
function pascalCase(...args: any[]): string {
  return args
    .filter(a => typeof a === 'string')
    .map(s => s[0].toUpperCase() + s.slice(1))
    .join('');
}

handlebars.registerHelper({ pascalCase });
export default handlebars;
