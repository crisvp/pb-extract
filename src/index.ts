export * from "./types.js";
export * from "./extract.js";
export * from "./sqlite.js";
export * from "./translations.js";
