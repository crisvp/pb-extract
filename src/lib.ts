import { Client as Pocketbase, type SchemaField } from "@crisvp/pocketbase-js";
import { writeFileSync } from "fs";
import path from "path";

const pascalCase = require("change-case").pascalCase;

export interface VitePluginPocketbaseOptions {
  dtsFile: string;
  jsonFile: string;
  authCollection: string;
  adminUser: string;
  adminPassword: string;
}

const DEFAULTS = {
  dtsFile: path.join(import.meta.dirname, "../src/pocketbase.d.ts"),
  jsonFile: path.join(import.meta.dirname, "../src/pocketbase.json"),

  authCollection: "_pb_users_auth_",
  adminUser: "vite@localhost.local",
  adminPassword: "SecretPassword123",
};

function tsType(field: SchemaField) {
  const { type } = field;
  switch (type) {
    case "text":
      return "string";
    case "date":
      return "Date";
    case "select":
      return field.options.values.map((v: string) => `'${v}'`).join(" | ");
    case "relation":
      return `%%relation:${field.options.collectionId}`;
    default:
      return "unknown";
  }
}

type Types = {
  [id: string]: {
    name: string;
    schema: Schema[];
  };
};

function schemaToString(collectionId: string, types: Types) {
  const { name, schema } = types[collectionId];
  return [
    `interface ${pascalCase(name)}Collection {`,
    ...schema.map((field) => {
      if (field.tsType.startsWith("%%relation:")) {
        const collectionId = field.tsType.replace("%%relation:", "");
        const collectionName = types[collectionId].name;
        return (collectionName ? `  ${field.name}: ${pascalCase(collectionName)}Collection` : "unknown") + ";";
      }
      return `  ${field.name}: ${field.tsType};`;
    }),
    "}",
  ];
}

export type Schema = {
  name: string;
  id: string;
  tsType: string;
};

export default function VitePluginPocketbase(options: Partial<VitePluginPocketbaseOptions> = {}) {
  return {
    name: "vite-plugin-pocketbase",
    async buildStart() {
      const pb = new Pocketbase("http://127.0.0.1:8090");
      const opts: VitePluginPocketbaseOptions = { ...DEFAULTS, ...options };
      const types: Types = {};

      await pb.admins.authWithPassword(opts.adminUser, opts.adminPassword);

      const collections = await pb.collections.getFullList();

      collections.map((collection) => {
        const schema = collection.schema.map((field) => ({
          name: field.name,
          id: field.id,
          tsType: tsType(field),
        }));
        types[collection.id] = { name: collection.name, schema };
        if (collection.type === "auth") {
          types[collection.id].schema.push(
            { name: "email", id: "", tsType: "string" },
            { name: "password", id: "", tsType: "string" },
            { name: "username", id: "", tsType: "string" },
          );
        }
      });

      const lines = collections.flatMap((collection) => schemaToString(collection.id, types)).map((str) => `  ${str}`);
      const interfaces = `export * from './pocketbase-auto';\n\ndeclare module './pocketbase-auto' {\n${lines.join(
        "\n",
      )}\n}\n`;

      const collectionOb = collections.map(
        (collection) => `    ${collection.name}: ${pascalCase(collection.name)}Collection;`,
      );
      const collectionMap = `declare module './pocketbase-auto' {\n  declare type Collections = {\n${collectionOb
        .map((s) => `  ${s}`)
        .join("\n")}\n  }\n}\n`;

      const prototypes = collections.flatMap((collection) => [
        `declare function mapCollection(name: '${collection.name}'): ${pascalCase(collection.name)}Collection;`,
      ]);

      const utility = "declare global {\n  declare type PocketbaseCollection<T> = UpperCase<`${T}Collection`>;\n}\n";

      const autoType = `declare module './pocketbase-auto' {\n${prototypes.map((s) => `  ${s}`).join("\n")}\n}\n`;
      const collectionNames = `declare module './pocketbase-auto' {\n  declare const CollectionNames = [${collections
        .map((collection) => `'${collection.name}'`)
        .join(",")}] as const;\n}\n`;

      const piniaLines = collections.map(
        (collection) => `  ${collection.name}: ${pascalCase(collection.name)}Collection;`,
      );
      const pinia = `export * from 'pinia';\ndeclare module 'pinia' {\n  export interface PiniaCustomProperties {\n    $pocketbase: {\n${piniaLines
        .map((s) => `    ${s}`)
        .join("\n")}\n    }\n  }\n}\n`;

      if (opts.dtsFile)
        writeFileSync(opts.dtsFile, [utility, interfaces, collectionMap, autoType, collectionNames, pinia].join("\n"));
      if (opts.jsonFile) writeFileSync(opts.jsonFile, JSON.stringify(types, null, 2));
    },
  };
}
