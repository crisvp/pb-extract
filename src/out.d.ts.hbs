{{! prettier-ignore }}
/** Interface definitions. These represent Pocketbase collections. */

{{! prettier-ignore }}
declare module 'pb-tex' {
{{#each tables}}

  export interface {{pascalCase this.name "collection"}} {
  {{#each this.schema}}
    {{this.name}}: {{{this.tsType}}};
  {{/each}}
  }
  {{/each}}
}
