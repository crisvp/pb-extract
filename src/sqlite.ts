import sqlite3 from "sqlite3";
import fs from "fs";
import path from "path";
import { CollectionDescriptionRaw, CollectionDescription } from "./types";
import { normalizeDescriptions } from "./translations.js";

/**
 * Open the database file.
 *
 * @param {string} fileName
 * @returns {sqlite3.Database}
 */
function openDatabase(fileName: string): sqlite3.Database {
  if (!fs.existsSync(fileName)) throw new Error(`File not found: ${fileName}`);
  return new sqlite3.Database(path.resolve(fileName), sqlite3.OPEN_READONLY);
}

/**
 * Reads the schema from a Pocketbase database. This uses the schema
 * as defined in Pocketbase, _not_ the schema as defined in SQLite.
 *
 * @param {sqlite3.Database} db Database object to extract types from
 * @returns {Promise<CollectionDescriptionRaw[]>}
 */
function collectionsTable(db: sqlite3.Database): Promise<CollectionDescriptionRaw[]> {
  return new Promise<CollectionDescriptionRaw[]>((resolve, reject) => {
    db.all("select id, type, name, schema from `_collections`", function (err, rows: CollectionDescriptionRaw[]) {
      if (err) reject(err);
      else resolve(rows);
    });
  });
}

/**
 * Reads a Pocketbase database file and returns normalized collection descriptions.
 *
 * @export
 * @async
 * @param {string} inputFile - Path to the input database file
 * @returns {Promise<CollectionDescription[]>}
 */
export async function readCollections(inputFile: string): Promise<CollectionDescription[]> {
  const db = openDatabase(inputFile);

  const schema = await collectionsTable(db);
  const extendedSchemas = normalizeDescriptions(schema);

  return extendedSchemas;
}
