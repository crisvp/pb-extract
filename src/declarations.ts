import fs from "fs";
import path from "path";

import handlebars from "./handlebars.js";
import { type CollectionDescription } from "./types";

export async function writeDeclarations(collections: CollectionDescription[], output: string): Promise<void> {
  const tpl = fs.readFileSync(path.join(import.meta.dirname, "out.d.ts.hbs"), "utf-8");
  const delegate = handlebars.compile(tpl);

  const outputStr = delegate({ tables: collections });
  fs.writeFileSync(path.resolve(output), outputStr);
}
