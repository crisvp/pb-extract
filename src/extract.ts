import path from "path";
import fs from "fs";
import handlebars from "./handlebars.js";
import { getSchema, openDatabase } from "./sqlite.js";
import { Schemas } from "./types.js";
import { addDefaultFields, addTsType, translateRelation } from "./translations.js";

/**
 * Reads a Pocketbase database file and outputs a d.ts file describing its collections.
 *
 * @export
 * @async
 * @param {string} inputFile - Path to the input database file
 * @returns {Promise<string>}
 */
export async function extractTypes(inputFile: string): Promise<string> {
  const db = openDatabase(inputFile);

  const tables = await getSchema(db);

  const extendedTables = tables.map((t) => addDefaultFields(t)).map((t) => addTsType(t));

  const schemas: Schemas = {};
  extendedTables.forEach((t) => (schemas[t.id] = { name: t.name, schema: t.schema }));
  extendedTables.forEach((t) => (t.schema = t.schema.map((s) => translateRelation(s, schemas))));

  const tpl = fs.readFileSync(path.join(import.meta.dirname, "out.d.ts.hbs"), "utf-8");
  const delegate = handlebars.compile(tpl);

  return delegate({ tables: schemas });
}
