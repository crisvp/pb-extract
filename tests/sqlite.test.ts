import { describe, expect, it } from 'vitest';
import { readCollections } from '../src/sqlite';

describe('sqlite', () => {
  it('should throw when file does not exist', async () => {
    expect(() => readCollections('asdf')).rejects.toThrowError('File not found: asdf');
  });

  it('should error when not a valid db', async () => {
    expect(() => readCollections('tests/fixtures/not-a-db.db')).rejects.toThrowError(/SQLITE_NOTADB/);
  });

  it('should read collections from db', async () => {
    const collections = await readCollections('tests/fixtures/data.db');
    expect(collections).toMatchSnapshot();
  });
});
