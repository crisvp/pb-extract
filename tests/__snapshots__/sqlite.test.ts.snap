// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sqlite > should read collections from db 1`] = `
[
  {
    "id": "_pb_users_auth_",
    "name": "users",
    "schema": [
      {
        "id": "unknown",
        "name": "id",
        "options": {},
        "presentable": true,
        "required": true,
        "system": false,
        "tsType": "string",
        "type": "text",
      },
      {
        "id": "unknown",
        "name": "created_at",
        "options": {},
        "presentable": true,
        "required": true,
        "system": false,
        "tsType": "Date",
        "type": "date",
      },
      {
        "id": "unknown",
        "name": "updated_at",
        "options": {},
        "presentable": true,
        "required": true,
        "system": false,
        "tsType": "Date",
        "type": "date",
      },
    ],
    "type": "auth",
  },
  {
    "id": "y7r8b764q36dm5k",
    "name": "relationships",
    "schema": [
      {
        "id": "unknown",
        "name": "id",
        "options": {},
        "presentable": true,
        "required": true,
        "system": false,
        "tsType": "string",
        "type": "text",
      },
      {
        "id": "unknown",
        "name": "created_at",
        "options": {},
        "presentable": true,
        "required": true,
        "system": false,
        "tsType": "Date",
        "type": "date",
      },
      {
        "id": "unknown",
        "name": "updated_at",
        "options": {},
        "presentable": true,
        "required": true,
        "system": false,
        "tsType": "Date",
        "type": "date",
      },
      {
        "id": "m4wle44g",
        "name": "userA",
        "options": {
          "cascadeDelete": false,
          "collectionId": "_pb_users_auth_",
          "displayFields": null,
          "maxSelect": 1,
          "minSelect": null,
        },
        "presentable": false,
        "required": false,
        "system": false,
        "tsType": "UsersCollection",
        "type": "relation",
        "unique": false,
      },
      {
        "id": "prhaed4e",
        "name": "userB",
        "options": {
          "cascadeDelete": false,
          "collectionId": "_pb_users_auth_",
          "displayFields": null,
          "maxSelect": 1,
          "minSelect": null,
        },
        "presentable": false,
        "required": false,
        "system": false,
        "tsType": "UsersCollection",
        "type": "relation",
        "unique": false,
      },
      {
        "id": "h27inftm",
        "name": "relationship",
        "options": {
          "maxSelect": 2,
          "values": [
            "friend",
            "blocked",
          ],
        },
        "presentable": false,
        "required": false,
        "system": false,
        "tsType": "'friend' | 'blocked'",
        "type": "select",
        "unique": false,
      },
    ],
    "type": "base",
  },
  {
    "id": "x9x94l3hve0wc8u",
    "name": "profile",
    "schema": [
      {
        "id": "unknown",
        "name": "id",
        "options": {},
        "presentable": true,
        "required": true,
        "system": false,
        "tsType": "string",
        "type": "text",
      },
      {
        "id": "unknown",
        "name": "created_at",
        "options": {},
        "presentable": true,
        "required": true,
        "system": false,
        "tsType": "Date",
        "type": "date",
      },
      {
        "id": "unknown",
        "name": "updated_at",
        "options": {},
        "presentable": true,
        "required": true,
        "system": false,
        "tsType": "Date",
        "type": "date",
      },
      {
        "id": "ssbrkau6",
        "name": "account",
        "options": {
          "cascadeDelete": false,
          "collectionId": "_pb_users_auth_",
          "displayFields": null,
          "maxSelect": 1,
          "minSelect": null,
        },
        "presentable": false,
        "required": true,
        "system": false,
        "tsType": "UsersCollection",
        "type": "relation",
        "unique": false,
      },
      {
        "id": "z0lrlsjp",
        "name": "real_name",
        "options": {
          "max": null,
          "min": null,
          "pattern": "",
        },
        "presentable": false,
        "required": false,
        "system": false,
        "tsType": "string",
        "type": "text",
        "unique": false,
      },
      {
        "id": "z1hea8x8",
        "name": "preferred_name",
        "options": {
          "max": null,
          "min": null,
          "pattern": "",
        },
        "presentable": false,
        "required": false,
        "system": false,
        "tsType": "string",
        "type": "text",
        "unique": false,
      },
      {
        "id": "ss5v5hmf",
        "name": "birthday",
        "options": {
          "max": "",
          "min": "",
        },
        "presentable": false,
        "required": false,
        "system": false,
        "tsType": "Date",
        "type": "date",
        "unique": false,
      },
    ],
    "type": "base",
  },
]
`;
